cli/cmd/upgrade.go

```diff
@@ -39,21 +39,37 @@ func newCmdUpgrade() *cobra.Command {
 	flags := options.recordableFlagSet()
 
 	cmd := &cobra.Command{
-		Use:   "upgrade [flags]",
-		Short: "Output Kubernetes configs to upgrade an existing Linkerd control plane",
+		Use:       fmt.Sprintf("upgrade [%s|%s] [flags]", configStage, controlPlaneStage),
+		Args:      cobra.OnlyValidArgs,
+		ValidArgs: []string{configStage, controlPlaneStage},
+		Short:     "Output Kubernetes configs to upgrade an existing Linkerd control plane",
 		Long: `Output Kubernetes configs to upgrade an existing Linkerd control plane.
 
 Note that the default flag values for this command come from the Linkerd control
 plane. The default values displayed in the Flags section below only apply to the
 install command.`,
+
+		Example: `  # Default upgrade.
+  linkerd upgrade | kubectl apply -f -
+
+  # Similar to install, upgrade may also be broken up into two stages, by user
+  # privilege.
+  # First stage requires cluster-level privileges.
+  linkerd upgrade config | kubectl apply -f -
+  # Second stage requires namespace-level privileges.
+  linkerd upgrade control-plane | kubectl apply -f -`,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			if options.ignoreCluster {
 				panic("ignore cluster must be unset") // Programmer error.
 			}
 
+			stage, err := validateArgs(args, flags, nil)
+			if err != nil {
+				return err
+			}
+
```

Comment:
Just noticed this now, when trying to test `--from-manifests` after @alpeb's comment. It seems, on line 79 of this file, that we are using a testing artifact to create a Kubernetes clientset. Is this intended? Wasn't clear on how its being used.

Reply:
Good question! (Also I forgot about @alpeb's original question re: allowing `linkerd config --from-manifests`, have pushed a new commit to prevent that.)

Re: `k8s.NewFakeClientSetsFromManifests` on line 79, this is intended. Adding `--from-manifests` required reading k8s yaml from stdin (or files), and processing them in lieu of k8s resources returned from the k8s api. We already had some test code doing this to mock out k8s clients. The change in #2697 moved that test code into a new `fake.go` file, to support `--from-manifests`. Leveraging `fake.go` allows most of the upgrade code to be agnostic of where the k8s resources came from, since it's just using a k8s client.